File: src/cryptic_nif.erl
%%% @doc Cryptic NIF Interface
%%%
%%% This module provides Erlang NIFs (Native Implemented Functions) for
%%% cryptographic operations used by the Cryptic chat application. It serves
%%% as the interface layer between Erlang code and the underlying C/C++
%%% cryptographic library implementations.
%%%
%%% == Features ==
%%%
%%% <ul>
%%%   <li>Curve25519 elliptic curve cryptography for key exchange</li>
%%%   <li>ChaCha20-Poly1305 authenticated encryption (AEAD)</li>
%%%   <li>Cryptographically secure random number generation</li>
%%%   <li>High-performance native cryptographic primitives</li>
%%%   <li>Memory-safe NIF implementations</li>
%%% </ul>
%%%
%%% == Cryptographic Operations ==
%%%
%%% The module provides the following cryptographic primitives:
%%% <ul>
%%%   <li>`gen_keypair/0' - Generate Curve25519 key pairs</li>
%%%   <li>`scalarmult/2' - Elliptic curve scalar multiplication (ECDH)</li>
%%%   <li>`aead_encrypt/3' - ChaCha20-Poly1305 authenticated encryption</li>
%%%   <li>`aead_decrypt/4' - ChaCha20-Poly1305 authenticated decryption</li>
%%%   <li>`rand_bytes/1' - Cryptographically secure random bytes</li>
%%% </ul>
%%%
%%% == Security Notes ==
%%%
%%% <ul>
%%%   <li>All cryptographic operations are performed in native code for performance</li>
%%%   <li>Memory is properly zeroed after use to prevent key material leakage</li>
%%%   <li>Random number generation uses the system's secure random source</li>
%%%   <li>Key material should be handled securely in calling code</li>
%%% </ul>
%%%
%%% == NIF Loading ==
%%%
%%% The module uses the `@on_load' attribute to automatically load the
%%% shared library containing the NIF implementations. The library is
%%% searched in standard priv directory locations.
%%%
%%% @author Cryptic Team
%%% @version 1.0.0
%%% @since 2025-09-14

-module(cryptic_nif).

-export([
    init/0,
    gen_keypair/0,
    scalarmult/2,
    aead_encrypt/3,
    aead_decrypt/4,
    rand_bytes/1,
    ed25519_sk_to_x25519_sk/1,
    ed25519_pk_to_x25519_pk/1,
    kdf_derive/4,
    hkdf_sha256/4,
    argon2id_raw/6
]).

-on_load(init/0).

-define(APPNAME, cryptic).
-define(LIBNAME, cryptic_nif).

%% @doc Initialize and load the NIF library
%%
%% This function is called automatically when the module is loaded due to
%% the `@on_load' attribute. It locates the shared library containing the
%% NIF implementations and loads it into the Erlang VM.
%%
%% == Library Search Process ==
%%
%% The function searches for the library in the following locations:
%% <ol>
%%%   <li>Application's priv directory (standard location)</li>
%%%   <li>Relative "../priv" directory (development builds)</li>
%%%   <li>Local "priv" directory (fallback)</li>
%% </ol>
%%
%% @returns ok if the NIF library loads successfully
%% @throws {error, any()}

init() ->
    SoName =
        case code:priv_dir(?APPNAME) of
            {error, bad_name} ->
                case filelib:is_dir(filename:join(["..", priv])) of
                    true ->
                        filename:join(["..", priv, ?LIBNAME]);
                    _ ->
                        filename:join([priv, ?LIBNAME])
                end;
            Dir ->
                filename:join(Dir, ?LIBNAME)
        end,
    erlang:load_nif(SoName, 0).

%% @doc Generate a Curve25519 key pair
%%
%% Generates a new Curve25519 elliptic curve key pair suitable for
%% Elliptic Curve Diffie-Hellman (ECDH) key exchange. The private key
%% should be kept secret while the public key can be shared.
%%
%% == Security Considerations ==
%%
%% <ul>
%%%   <li>Private keys must be kept secure and never transmitted</li>
%%%   <li>Public keys can be safely shared for key exchange</li>
%%%   <li>Keys are generated using cryptographically secure randomness</li>
%%%   <li>Private key material is properly handled in native code</li>
%% </ul>
%%
%% @returns {PublicKey, PrivateKey} where both are 32-byte binaries
%% @throws {error, atom()} | {nif_not_loaded, atom()}
gen_keypair() ->
    erlang:nif_error({nif_not_loaded, ?MODULE}).

%% @doc Perform Curve25519 scalar multiplication (ECDH)
%%
%% Computes the shared secret between a private key and a public key
%% using Curve25519 elliptic curve scalar multiplication. This is the
%% core operation for Elliptic Curve Diffie-Hellman key exchange.
%%
%% == Usage Pattern ==
%%
%% ```
%% % Alice generates her key pair
%% {AlicePub, AlicePriv} = cryptic_nif:gen_keypair(),
%%
%% % Bob generates his key pair
%% {BobPub, BobPriv} = cryptic_nif:gen_keypair(),
%%
%% % Both parties compute the same shared secret
%% SharedSecret1 = cryptic_nif:scalarmult(AlicePriv, BobPub),
%% SharedSecret2 = cryptic_nif:scalarmult(BobPriv, AlicePub),
%% % SharedSecret1 =:= SharedSecret2
%% '''
%%
%% == Security Notes ==
%%
%% <ul>
%%%   <li>The shared secret should be used as input to a key derivation function</li>
%%%   <li>Never use the shared secret directly as an encryption key</li>
%%%   <li>Private keys must be exactly 32 bytes</li>
%%%   <li>Public keys must be exactly 32 bytes and on the curve</li>
%% </ul>
%%
%% @param SecretKey The private key (32-byte binary)
%% @param PublicKey The peer's public key (32-byte binary)
%% @returns SharedSecret A 32-byte binary containing the shared secret
%% @throws {error, atom()} | {nif_not_loaded, atom()}

scalarmult(_SecretKey, _PublicKey) ->
    erlang:nif_error({nif_not_loaded, ?MODULE}).

%% @doc Encrypt data using ChaCha20-Poly1305 AEAD
%%
%% Performs authenticated encryption using the ChaCha20-Poly1305 AEAD
%% (Authenticated Encryption with Associated Data) algorithm. This provides
%% both confidentiality and authenticity for the encrypted data.
%%
%% == Algorithm Details ==
%%
%% <ul>
%%%   <li>Encryption: ChaCha20 stream cipher</li>
%%%   <li>Authentication: Poly1305 MAC</li>
%%%   <li>Nonce: 12-byte random value (generated automatically)</li>
%%%   <li>Key: 32-byte encryption key</li>
%%%   <li>AAD: Additional authenticated data (not encrypted)</li>
%% </ul>
%%
%% == Usage Example ==
%%
%% ```
%% Key = cryptic_nif:rand_bytes(32),
%% Plaintext = <<"Hello, World!">>,
%% AAD = <<"metadata">>,
%% {Nonce, Ciphertext} = cryptic_nif:aead_encrypt(Plaintext, Key, AAD).
%% '''
%%
%% == Security Notes ==
%%
%% <ul>
%%%   <li>Never reuse the same key+nonce combination</li>
%%%   <li>Nonce is generated randomly and must be transmitted with ciphertext</li>
%%%   <li>AAD is authenticated but not encrypted</li>
%%%   <li>Tampering with ciphertext or AAD will cause decryption to fail</li>
%% </ul>
%%
%% @param Plaintext The data to encrypt (binary)
%% @param Key The 32-byte encryption key (binary)
%% @param AAD Additional authenticated data (binary, can be empty)
%% @returns {Nonce, Ciphertext} where Nonce is 12 bytes and Ciphertext includes auth tag
%% @throws {error, atom()} | {nif_not_loaded, atom()}

aead_encrypt(_Plaintext, _Key, _AAD) ->
    erlang:nif_error({nif_not_loaded, ?MODULE}).

%% @doc Decrypt data using ChaCha20-Poly1305 AEAD
%%
%% Performs authenticated decryption using the ChaCha20-Poly1305 AEAD
%% algorithm. Verifies both the authenticity and integrity of the encrypted
%% data before returning the plaintext.
%%
%% == Verification Process ==
%%
%% <ol>
%%%   <li>Verify the Poly1305 authentication tag</li>
%%%   <li>Verify the AAD has not been tampered with</li>
%%%   <li>Decrypt the ciphertext using ChaCha20</li>
%%%   <li>Return plaintext only if all verifications pass</li>
%% </ol>
%%
%% == Usage Example ==
%%
%% ```
%% % Using data from aead_encrypt/3
%% case cryptic_nif:aead_decrypt(Ciphertext, Key, Nonce, AAD) of
%%     {ok, Plaintext} ->
%%         % Decryption successful, data is authentic
%%         process_plaintext(Plaintext);
%%     {error, auth_failed} ->
%%         % Data has been tampered with or wrong key/nonce
%%         handle_auth_failure()
%% end.
%% '''
%%
%% == Security Notes ==
%%
%% <ul>
%%%   <li>Always check the return value - authentication failure indicates tampering</li>
%%%   <li>Use the exact same key, nonce, and AAD that were used for encryption</li>
%%%   <li>Never attempt to process data if authentication fails</li>
%%%   <li>Timing attacks are mitigated by constant-time verification</li>
%% </ul>
%%
%% @param Ciphertext The encrypted data including authentication tag (binary)
%% @param Key The 32-byte decryption key (binary)
%% @param Nonce The 12-byte nonce used during encryption (binary)
%% @param AAD The additional authenticated data (binary, must match encryption)
%% @returns {ok, Plaintext} on successful decryption and authentication,
%%          {error, auth_failed} if authentication fails
%% @throws {error, atom()} | {nif_not_loaded, atom()}

aead_decrypt(_Ciphertext, _Key, _Nonce, _AAD) ->
    erlang:nif_error({nif_not_loaded, ?MODULE}).

%% @doc Generate cryptographically secure random bytes
%%
%% Generates the specified number of cryptographically secure random bytes
%% using the system's secure random number generator. This function is
%% suitable for generating keys, nonces, salts, and other security-critical
%% random data.
%%
%% == Random Source ==
%%
%% <ul>
%%%   <li>Uses the operating system's secure random number generator</li>
%%%   <li>On Linux: /dev/urandom or getrandom() syscall</li>
%%%   <li>On macOS: SecRandomCopyBytes or /dev/urandom</li>
%%%   <li>On Windows: CryptGenRandom or BCryptGenRandom</li>
%% </ul>
%%
%% == Usage Examples ==
%%
%% ```
%% % Generate a 32-byte encryption key
%% Key = cryptic_nif:rand_bytes(32),
%%
%% % Generate a 12-byte nonce
%% Nonce = cryptic_nif:rand_bytes(12),
%%
%% % Generate a random salt
%% Salt = cryptic_nif:rand_bytes(16).
%% '''
%%
%% == Security Notes ==
%%
%% <ul>
%%%   <li>Output is suitable for cryptographic purposes</li>
%%%   <li>No need to seed - uses system entropy</li>
%%%   <li>Non-blocking operation in most cases</li>
%%%   <li>Returns different values on each call</li>
%% </ul>
%%
%% @param Size The number of random bytes to generate (positive integer)
%% @returns Binary containing the requested number of random bytes
%% @throws {error, atom()} | {nif_not_loaded, atom()}

rand_bytes(_Size) ->
    erlang:nif_error({nif_not_loaded, ?MODULE}).

%% @doc Convert Ed25519 private key to X25519 private key
%%
%% Performs a secure conversion from an Ed25519 signing private key to an
%% X25519 ECDH private key using libsodium's crypto_sign_ed25519_sk_to_curve25519.
%% This conversion is mathematically sound and preserves the cryptographic
%% relationship between the keys.
%%
%% == Usage ==
%%
%% ```
%% {Ed25519Pub, Ed25519Priv} = crypto:generate_key(eddsa, ed25519),
%% X25519Priv = cryptic_nif:ed25519_sk_to_x25519_sk(Ed25519Priv).
%% '''
%%
%% == Security Notes ==
%%
%% <ul>
%%%   <li>The conversion is deterministic - same Ed25519 key produces same X25519 key</li>
%%%   <li>The converted key is cryptographically valid for X25519 operations</li>
%%%   <li>Both keys should be treated with equal security precautions</li>
%% </ul>
%%
%% @param Ed25519SecretKey The Ed25519 private key (32-byte binary)
%% @returns X25519 private key (32-byte binary)
%% @throws {error, atom()} | {nif_not_loaded, atom()}
ed25519_sk_to_x25519_sk(_Ed25519SecretKey) ->
    erlang:nif_error({nif_not_loaded, ?MODULE}).

%% @doc Convert Ed25519 public key to X25519 public key
%%
%% Performs a secure conversion from an Ed25519 signing public key to an
%% X25519 ECDH public key using libsodium's crypto_sign_ed25519_pk_to_curve25519.
%% This conversion maintains the mathematical relationship between the corresponding
%% private keys.
%%
%% == Usage ==
%%
%% ```
%% {Ed25519Pub, Ed25519Priv} = crypto:generate_key(eddsa, ed25519),
%% X25519Pub = cryptic_nif:ed25519_pk_to_x25519_pk(Ed25519Pub).
%% '''
%%
%% == Security Notes ==
%%
%% <ul>
%%%   <li>The conversion is deterministic and publicly computable</li>
%%%   <li>Both public keys can be safely shared</li>
%%%   <li>The converted X25519 key corresponds to the converted private key</li>
%% </ul>
%%
%% @param Ed25519PublicKey The Ed25519 public key (32-byte binary)
%% @returns X25519 public key (32-byte binary)
%% @throws {error, atom()} | {nif_not_loaded, atom()}
ed25519_pk_to_x25519_pk(_Ed25519PublicKey) ->
    erlang:nif_error({nif_not_loaded, ?MODULE}).

%% @doc Derive key material using libsodium's KDF (Blake2b-based)
%%
%% High-performance key derivation using libsodium's crypto_kdf_derive_from_key
%% function. This is optimized for symmetric key ratcheting where many keys
%% need to be derived quickly from a master key.
%%
%% == Ratcheting Usage ==
%%
%% ```
%% % Initialize with root key
%% RootKey = cryptic_nif:rand_bytes(32),
%% Context = <<"ratchet">>,
%%
%% % Derive chain keys with incrementing IDs
%% ChainKey1 = cryptic_nif:kdf_derive(32, 1, Context, RootKey),
%% ChainKey2 = cryptic_nif:kdf_derive(32, 2, Context, RootKey),
%% MessageKey1 = cryptic_nif:kdf_derive(32, 1, <<"msg">>, ChainKey1).
%% '''
%%
%% == Performance Notes ==
%%
%% <ul>
%%%   <li>Faster than HKDF-SHA256 for high-frequency operations</li>
%%%   <li>Uses Blake2b internally which is optimized for speed</li>
%%%   <li>Perfect for Double Ratchet message key derivation</li>
%%%   <li>Constant-time operation prevents timing attacks</li>
%% </ul>
%%
%% == Security Properties ==
%%
%% <ul>
%%%   <li>Cryptographically secure key separation by SubkeyId</li>
%%%   <li>Context provides domain separation (max 8 bytes)</li>
%%%   <li>Different SubkeyId values produce independent keys</li>
%%%   <li>Cannot derive master key from any derived keys</li>
%% </ul>
%%
%% @param Length Length of derived key in bytes (1-64)
%% @param SubkeyId Unique identifier for this derived key (0-2^64-1)
%% @param Context Domain separation context (max 8 bytes)
%% @param MasterKey Master key material (32 bytes)
%% @returns Derived key binary of specified length
%% @throws {error, atom()} | {nif_not_loaded, atom()}
kdf_derive(_Length, _SubkeyId, _Context, _MasterKey) ->
    erlang:nif_error({nif_not_loaded, ?MODULE}).

%% @doc HKDF-SHA256 key derivation (RFC 5869 compliant)
%%
%% Standard HKDF-SHA256 implementation for compatibility with existing
%% protocols and systems that specifically require RFC 5869 HKDF.
%% For new applications, consider kdf_derive/4 which is faster.
%%
%% == Protocol Compatibility ==
%%
%% ```
%% % X3DH shared secret expansion
%% SharedSecret = cryptic_nif:scalarmult(PrivKey, PubKey),
%% Salt = <<"X3DH">>,
%% Info = <<"encryption_key">>,
%% EncKey = cryptic_nif:hkdf_sha256(SharedSecret, Salt, Info, 32).
%% '''
%%
%% == RFC 5869 Process ==
%%
%% <ol>
%%%   <li>Extract: PRK = HMAC-SHA256(Salt, IKM)</li>
%%%   <li>Expand: OKM = HMAC-SHA256(PRK, Info || Counter)</li>
%% </ol>
%%
%% == Security Notes ==
%%
%% <ul>
%%%   <li>Salt should be random or at least unique per protocol run</li>
%%%   <li>Info provides context and domain separation</li>
%%%   <li>Same IKM+Salt+Info always produces same output (deterministic)</li>
%%%   <li>Current implementation supports up to 32-byte output</li>
%% </ul>
%%
%% @param IKM Input keying material (e.g., ECDH shared secret)
%% @param Salt Salt value for extraction phase (can be an empty binary)
%% @param Info Context and application-specific information
%% @param Length Length of output keying material (1-32 bytes)
%% @returns Derived key material of specified length
%% @throws {error, atom()} | {nif_not_loaded, atom()}
hkdf_sha256(_IKM, _Salt, _Info, _Length) ->
    erlang:nif_error({nif_not_loaded, ?MODULE}).

%% @doc Derive raw key material using Argon2id (libargon2)
%%
%% Computes an Argon2id (version 1.3) hash with a fully configurable parameter
%% set. Unlike a general password-hashing helper, this exposes `Parallelism'
%% because the enrollment package format produced by `bin/cryptic-onboard'
%% uses the `argon2' CLI with a lane count of 4, which the libsodium
%% high-level `crypto_pwhash' API cannot reproduce.
%%
%% == Usage ==
%%
%% ```
%% %% Mirrors: argon2 <Salt> -id -t 3 -m 16 -p 4 -l 64 -r
%% Derived = cryptic_nif:argon2id_raw(Passphrase, Salt, 3, 65536, 4, 64).
%% '''
%%
%% == Parameters ==
%%
%% <ul>
%%   <li>`MemoryKiB' is the memory cost in kibibytes. The CLI `-m N' flag
%%       corresponds to `2^N' KiB, so `-m 16' means pass `65536'.</li>
%%   <li>`Salt' is used verbatim as the Argon2 salt (must be &gt;= 8 bytes).</li>
%% </ul>
%%
%% @param Password The secret input (binary)
%% @param Salt The salt bytes (binary, at least 8 bytes)
%% @param TimeCost Number of iterations (&gt;= 1)
%% @param MemoryKiB Memory cost in kibibytes
%% @param Parallelism Number of lanes/threads (&gt;= 1)
%% @param HashLen Length of the derived output in bytes (4-1024)
%% @returns Raw derived key material of `HashLen' bytes
%% @throws {error, atom()} | {nif_not_loaded, atom()}
argon2id_raw(_Password, _Salt, _TimeCost, _MemoryKiB, _Parallelism, _HashLen) ->
    erlang:nif_error({nif_not_loaded, ?MODULE}).

File: c_src/cryptic_nif.c
/**
 * @file cryptic_nif.c
 * @brief Cryptographic NIF (Native Implemented Functions) for Cryptic Chat Application
 *
 * This module provides high-performance cryptographic primitives through libsodium
 * for secure end-to-end encrypted messaging. It implements essential cryptographic
 * operations required for the X3DH key agreement protocol and ChaCha20-Poly1305
 * authenticated encryption.
 *
 * @section Features
 * - X25519 elliptic curve key generation and scalar multiplication
 * - ChaCha20-Poly1305 AEAD encryption and decryption
 * - Ed25519 to X25519 key conversion for dual-key cryptography
 * - Cryptographically secure random number generation
 * - Memory-safe operations with automatic cleanup of sensitive data
 *
 * @section Security
 * All functions implement proper key clamping, memory zeroing, and error handling
 * to prevent timing attacks and sensitive data leakage.
 *
 * @author Cryptic Team
 * @version 2.0
 * @date 2025-09-19
 * @copyright MIT License
 */

#include <erl_nif.h>
#include <sodium.h>
#include <string.h>

// NIF resource types
static ErlNifResourceType *keypair_type = NULL;

/**
 * @brief Initialize the NIF module and libsodium
 *
 * This function is called when the NIF library is loaded. It initializes
 * libsodium's secure random number generator and sets up Erlang resource types.
 *
 * @param env Erlang NIF environment
 * @param priv_data Private data pointer (unused)
 * @param load_info Load information term (unused)
 * @return 0 on success, -1 on failure
 */
static int load(ErlNifEnv *env, void **priv_data, ERL_NIF_TERM load_info)
{
    if (sodium_init() < 0)
    {
        return -1;
    }

    keypair_type = enif_open_resource_type(env, NULL, "keypair", NULL,
                                           ERL_NIF_RT_CREATE, NULL);
    if (keypair_type == NULL)
    {
        return -1;
    }

    return 0;
}

/**
 * @brief Generate an X25519 elliptic curve keypair
 *
 * Generates a cryptographically secure X25519 keypair suitable for Elliptic Curve
 * Diffie-Hellman (ECDH) key agreement. The private key is properly clamped according
 * to RFC 7748 specifications to ensure security.
 *
 * @param env Erlang NIF environment
 * @param argc Number of arguments (must be 0)
 * @param argv Array of arguments (unused)
 * @return Tuple {PublicKey, PrivateKey} or atom 'error'
 *
 * @note The private key is automatically zeroed from memory after copying to prevent
 *       sensitive data leakage.
 */
static ERL_NIF_TERM gen_keypair(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
{
    unsigned char pk[crypto_scalarmult_BYTES];
    unsigned char sk[crypto_scalarmult_SCALARBYTES];

    // Generate random private key
    randombytes_buf(sk, crypto_scalarmult_SCALARBYTES);

    // Clamp the private key for X25519 (this is crucial!)
    sk[0] &= 248;
    sk[31] &= 127;
    sk[31] |= 64;

    // Compute public key from private key
    if (crypto_scalarmult_base(pk, sk) != 0)
    {
        return enif_make_atom(env, "error");
    }

    ERL_NIF_TERM public_key, secret_key;
    unsigned char *pk_data = enif_make_new_binary(env, crypto_scalarmult_BYTES, &public_key);
    unsigned char *sk_data = enif_make_new_binary(env, crypto_scalarmult_SCALARBYTES, &secret_key);

    memcpy(pk_data, pk, crypto_scalarmult_BYTES);
    memcpy(sk_data, sk, crypto_scalarmult_SCALARBYTES);

    // Clear sensitive data
    sodium_memzero(sk, crypto_scalarmult_SCALARBYTES);

    return enif_make_tuple2(env, public_key, secret_key);
}

/**
 * @brief Perform X25519 scalar multiplication for shared secret derivation
 *
 * Computes the shared secret between a private key and a public key using
 * X25519 elliptic curve scalar multiplication. This is the core operation
 * for ECDH key agreement.
 *
 * @param env Erlang NIF environment
 * @param argc Number of arguments (must be 2)
 * @param argv Array containing [PrivateKey, PublicKey] binaries
 * @return 32-byte shared secret binary or atom 'error'/'badarg'
 *
 * @security The shared secret is zeroed from memory after copying to prevent
 *           sensitive data leakage.
 */
static ERL_NIF_TERM scalarmult(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
{
    ErlNifBinary secret_key, public_key;

    if (!enif_inspect_binary(env, argv[0], &secret_key) ||
        !enif_inspect_binary(env, argv[1], &public_key) ||
        secret_key.size != crypto_scalarmult_SCALARBYTES ||
        public_key.size != crypto_scalarmult_BYTES)
    {
        return enif_make_badarg(env);
    }

    unsigned char shared_secret[crypto_scalarmult_BYTES];

    if (crypto_scalarmult(shared_secret, secret_key.data, public_key.data) != 0)
    {
        return enif_make_atom(env, "error");
    }

    ERL_NIF_TERM result;
    unsigned char *result_data = enif_make_new_binary(env, crypto_scalarmult_BYTES, &result);
    memcpy(result_data, shared_secret, crypto_scalarmult_BYTES);

    // Clear sensitive data
    sodium_memzero(shared_secret, crypto_scalarmult_BYTES);

    return result;
}

/**
 * @brief Encrypt plaintext using ChaCha20-Poly1305 AEAD cipher
 *
 * Performs authenticated encryption with associated data (AEAD) using the
 * ChaCha20-Poly1305 construction. This provides both confidentiality and
 * authenticity guarantees.
 *
 * @param env Erlang NIF environment
 * @param argc Number of arguments (must be 3)
 * @param argv Array containing [Plaintext, Key, AAD] binaries
 * @return Tuple {Ciphertext, Nonce} or atom 'error'/'badarg'
 *
 * @note The nonce is randomly generated for each encryption operation.
 *       The key must be exactly 32 bytes (256 bits).
 * @security Uses cryptographically secure random nonce generation.
 */
static ERL_NIF_TERM aead_encrypt(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
{
    ErlNifBinary plaintext, key, aad;

    if (!enif_inspect_binary(env, argv[0], &plaintext) ||
        !enif_inspect_binary(env, argv[1], &key) ||
        !enif_inspect_binary(env, argv[2], &aad) ||
        key.size != crypto_aead_chacha20poly1305_ietf_KEYBYTES)
    {
        return enif_make_badarg(env);
    }

    unsigned char nonce[crypto_aead_chacha20poly1305_ietf_NPUBBYTES];
    unsigned char *ciphertext = malloc(plaintext.size + crypto_aead_chacha20poly1305_ietf_ABYTES);
    unsigned long long ciphertext_len;

    if (ciphertext == NULL)
    {
        return enif_make_atom(env, "error");
    }

    // Generate random nonce
    randombytes_buf(nonce, sizeof(nonce));

    if (crypto_aead_chacha20poly1305_ietf_encrypt(
            ciphertext, &ciphertext_len,
            plaintext.data, plaintext.size,
            aad.data, aad.size,
            NULL, nonce, key.data) != 0)
    {
        free(ciphertext);
        return enif_make_atom(env, "error");
    }

    ERL_NIF_TERM cipher_term, nonce_term;
    unsigned char *cipher_data = enif_make_new_binary(env, ciphertext_len, &cipher_term);
    unsigned char *nonce_data = enif_make_new_binary(env, sizeof(nonce), &nonce_term);

    memcpy(cipher_data, ciphertext, ciphertext_len);
    memcpy(nonce_data, nonce, sizeof(nonce));

    free(ciphertext);

    return enif_make_tuple2(env, cipher_term, nonce_term);
}

/**
 * @brief Decrypt ciphertext using ChaCha20-Poly1305 AEAD cipher
 *
 * Performs authenticated decryption with associated data (AEAD) verification
 * using ChaCha20-Poly1305. Verifies both authenticity and integrity before
 * returning plaintext.
 *
 * @param env Erlang NIF environment
 * @param argc Number of arguments (must be 4)
 * @param argv Array containing [Ciphertext, Key, Nonce, AAD] binaries
 * @return Plaintext binary or atom 'error'/'badarg'
 *
 * @note The key must be exactly 32 bytes and nonce exactly 12 bytes.
 *       Ciphertext must be at least 16 bytes (minimum tag size).
 * @security Authentication tag is verified before returning plaintext.
 *           Returns 'error' if authentication fails.
 */
static ERL_NIF_TERM aead_decrypt(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
{
    ErlNifBinary ciphertext, key, nonce, aad;

    if (!enif_inspect_binary(env, argv[0], &ciphertext) ||
        !enif_inspect_binary(env, argv[1], &key) ||
        !enif_inspect_binary(env, argv[2], &nonce) ||
        !enif_inspect_binary(env, argv[3], &aad) ||
        key.size != crypto_aead_chacha20poly1305_ietf_KEYBYTES ||
        nonce.size != crypto_aead_chacha20poly1305_ietf_NPUBBYTES)
    {
        return enif_make_badarg(env);
    }

    if (ciphertext.size < crypto_aead_chacha20poly1305_ietf_ABYTES)
    {
        return enif_make_atom(env, "error");
    }

    unsigned char *plaintext = malloc(ciphertext.size - crypto_aead_chacha20poly1305_ietf_ABYTES);
    unsigned long long plaintext_len;

    if (plaintext == NULL)
    {
        return enif_make_atom(env, "error");
    }

    if (crypto_aead_chacha20poly1305_ietf_decrypt(
            plaintext, &plaintext_len,
            NULL,
            ciphertext.data, ciphertext.size,
            aad.data, aad.size,
            nonce.data, key.data) != 0)
    {
        free(plaintext);
        return enif_make_atom(env, "error");
    }

    ERL_NIF_TERM result;
    unsigned char *result_data = enif_make_new_binary(env, plaintext_len, &result);
    memcpy(result_data, plaintext, plaintext_len);

    free(plaintext);

    return result;
}

/**
 * @brief Generate cryptographically secure random bytes
 *
 * Generates random bytes suitable for cryptographic use, including keys,
 * nonces, salts, and other security-critical values.
 *
 * @param env Erlang NIF environment
 * @param argc Number of arguments (must be 1)
 * @param argv Array containing [Size] integer
 * @return Binary containing random bytes or atom 'badarg'
 *
 * @note Size is limited to 1024 bytes to prevent excessive memory allocation.
 * @security Uses libsodium's cryptographically secure PRNG.
 */
static ERL_NIF_TERM rand_bytes(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
{
    int size;

    if (!enif_get_int(env, argv[0], &size) || size < 0 || size > 1024)
    {
        return enif_make_badarg(env);
    }

    ERL_NIF_TERM result;
    unsigned char *data = enif_make_new_binary(env, size, &result);
    randombytes_buf(data, size);

    return result;
}

/**
 * @brief Convert Ed25519 private key (seed) to X25519 private key
 *
 * Converts an Ed25519 signing key to an X25519 key agreement key, enabling
 * dual-key cryptography where the same identity can be used for both
 * signatures and key agreement.
 *
 * @param env Erlang NIF environment
 * @param argc Number of arguments (must be 1)
 * @param argv Array containing [Ed25519Seed] binary (32 bytes)
 * @return X25519 private key binary (32 bytes) or atom 'error'/'badarg'
 *
 * @note Input should be Ed25519 seed (32 bytes), not full secretkey (64 bytes).
 *       This matches Erlang's crypto:generate_key(eddsa, ed25519) output.
 * @security All intermediate Ed25519 key material is zeroed after conversion.
 */
static ERL_NIF_TERM ed25519_sk_to_x25519_sk(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
{
    ErlNifBinary ed25519_seed;

    // Erlang's crypto:generate_key(eddsa, ed25519) returns 32-byte seed, not 64-byte secretkey
    if (!enif_inspect_binary(env, argv[0], &ed25519_seed) ||
        ed25519_seed.size != crypto_sign_SEEDBYTES)
    {
        return enif_make_badarg(env);
    }

    // First expand the seed to a full Ed25519 keypair
    unsigned char ed25519_pk[crypto_sign_PUBLICKEYBYTES];
    unsigned char ed25519_sk[crypto_sign_SECRETKEYBYTES];

    if (crypto_sign_seed_keypair(ed25519_pk, ed25519_sk, ed25519_seed.data) != 0)
    {
        return enif_make_atom(env, "error");
    }

    ERL_NIF_TERM result;
    unsigned char *x25519_sk = enif_make_new_binary(env, crypto_scalarmult_SCALARBYTES, &result);

    if (crypto_sign_ed25519_sk_to_curve25519(x25519_sk, ed25519_sk) != 0)
    {
        // Clear sensitive data
        sodium_memzero(ed25519_sk, crypto_sign_SECRETKEYBYTES);
        return enif_make_atom(env, "error");
    }

    // Clear sensitive data
    sodium_memzero(ed25519_sk, crypto_sign_SECRETKEYBYTES);

    return result;
}

/**
 * @brief Convert Ed25519 public key to X25519 public key
 *
 * Converts an Ed25519 verification key to an X25519 key agreement key,
 * completing the dual-key cryptography setup for public key operations.
 *
 * @param env Erlang NIF environment
 * @param argc Number of arguments (must be 1)
 * @param argv Array containing [Ed25519PublicKey] binary (32 bytes)
 * @return X25519 public key binary (32 bytes) or atom 'error'/'badarg'
 *
 * @note This is safe to perform on public keys as no sensitive data is involved.
 *       Essential for X3DH protocol implementation.
 */
static ERL_NIF_TERM ed25519_pk_to_x25519_pk(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
{
    ErlNifBinary ed25519_pk;

    if (!enif_inspect_binary(env, argv[0], &ed25519_pk) ||
        ed25519_pk.size != crypto_sign_PUBLICKEYBYTES)
    {
        return enif_make_badarg(env);
    }

    ERL_NIF_TERM result;
    unsigned char *x25519_pk = enif_make_new_binary(env, crypto_scalarmult_BYTES, &result);

    if (crypto_sign_ed25519_pk_to_curve25519(x25519_pk, ed25519_pk.data) != 0)
    {
        return enif_make_atom(env, "error");
    }

    return result;
}

/**
 * @brief NIF function export table
 *
 * Defines the mapping between Erlang function names and C implementations.
 * All functions are marked as dirty CPU-bound operations for proper scheduling.
 */
static ErlNifFunc nif_funcs[] = {
    {"gen_keypair", 0, gen_keypair, 0},
    {"scalarmult", 2, scalarmult, 0},
    {"aead_encrypt", 3, aead_encrypt, 0},
    {"aead_decrypt", 4, aead_decrypt, 0},
    {"rand_bytes", 1, rand_bytes, 0},
    {"ed25519_sk_to_x25519_sk", 1, ed25519_sk_to_x25519_sk, 0},
    {"ed25519_pk_to_x25519_pk", 1, ed25519_pk_to_x25519_pk, 0}};

/**
 * @brief NIF module initialization macro
 *
 * Registers the NIF module with the Erlang runtime system.
 * Specifies load/unload callbacks and function table.
 */
ERL_NIF_INIT(cryptic_nif, nif_funcs, load, NULL, NULL, NULL)
